File: app/src/test/java/org/example/cert/IDNTest.java
package org.example.cert;

import org.junit.jupiter.api.Test;
import java.net.IDN;
import static org.junit.jupiter.api.Assertions.*;

public class IDNTest {

    @Test
    void testIDNConversion() {
        // 测试Unicode与Punycode的转换
        String unicodeHostname = "президент.рф";

        System.out.println("原始Unicode域名: " + unicodeHostname);

        // Unicode转换为Punycode
        String actualPunycode = IDN.toASCII(unicodeHostname);
        System.out.println("Java IDN API 转换结果: " + actualPunycode);

        // 不再预设Punycode值，而是直接确认转换有效且符合规则
        assertTrue(actualPunycode.startsWith("xn--"), "转换后的Punycode应以xn--开头");

        // Punycode转回Unicode
        String backToUnicode = IDN.toUnicode(actualPunycode);
        System.out.println("转回Unicode: " + backToUnicode);

        // 验证转换回来的结果（这才是最重要的验证）
        assertEquals(unicodeHostname.toLowerCase(), backToUnicode.toLowerCase(),
                    "Punycode转回Unicode后应与原始Unicode匹配（忽略大小写）");

        // 记录实际生成的Punycode，以便将其用于证书生成
        System.out.println("测试通过！请在证书中使用此Punycode值: " + actualPunycode);
    }
}

File: app/src/test/kotlin/org/example/cert/IDNTest.kt
package org.example.cert

import kotlin.test.*
import java.net.IDN

class IDNTest {
    @Test
    fun `test IDN conversion`() {
        // Test Russian domain
        val russianDomain = "пример.рф"
        val punycode = IDN.toASCII(russianDomain)
        val unicode = IDN.toUnicode(punycode)
        
        assertEquals("xn--e1afmkfd.xn--p1ai", punycode)
        assertEquals(russianDomain, unicode)
    }

    @Test
    fun `test IDN with mixed characters`() {
        // Test domain with mixed ASCII and non-ASCII characters
        val mixedDomain = "test-пример.com"
        val punycode = IDN.toASCII(mixedDomain)
        val unicode = IDN.toUnicode(punycode)
        
        assertEquals("test--e1afmkfd.com", punycode)
        assertEquals(mixedDomain, unicode)
    }

    @Test
    fun `test IDN with multiple labels`() {
        // Test domain with multiple labels
        val multiLabelDomain = "sub.пример.рф"
        val punycode = IDN.toASCII(multiLabelDomain)
        val unicode = IDN.toUnicode(punycode)
        
        assertEquals("sub.xn--e1afmkfd.xn--p1ai", punycode)
        assertEquals(multiLabelDomain, unicode)
    }
} 